File: java/backend-spring/src/main/java/com/countriesdb/examples/spring/ValidationController02.java
package com.countriesdb.examples.spring;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
@RequestMapping("/api/validate")
public class ValidationController02 {
    private final CountriesDbClient client;

    public ValidationController02(CountriesDbClient client) {
        this.client = client;
    }

    // 2. Country validation with follow_upward
    @PostMapping("/country/follow-upward")
    public ResponseEntity<?> validateCountryFollowUpward(@RequestBody Map<String, String> payload) {
        String country = payload.get("country");
        if (country == null || country.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of(
                "valid", false,
                "message", "Country code is required"
            ));
        }
        
        Map<String, Object> result = client.validateCountry(Map.of(
            "code", country,
            "follow_upward", true
        ));
        
        if (!Boolean.TRUE.equals(result.get("valid"))) {
            return ResponseEntity.unprocessableEntity().body(result);
        }
        
        return ResponseEntity.ok(Map.of(
            "valid", true,
            "message", "Country validated",
            "payload", payload
        ));
    }
}

File: csharp/backend-aspnetcore/Controllers/ValidationController02.cs
using Microsoft.AspNetCore.Mvc;

namespace BackendAspNetCore.Controllers;

[ApiController]
[Route("api/validate")]
public class ValidationController02 : ControllerBase
{
    private readonly CountriesDbClient _client;

    public ValidationController02(CountriesDbClient client)
    {
        _client = client;
    }

    // 2. Country validation with follow_upward
    [HttpPost("country/follow-upward")]
    public async Task<IActionResult> ValidateCountryFollowUpward([FromBody] Dictionary<string, string> payload)
    {
        var country = payload.GetValueOrDefault("country");
        if (string.IsNullOrWhiteSpace(country))
        {
            return BadRequest(new Dictionary<string, object>
            {
                ["valid"] = false,
                ["message"] = "Country code is required"
            });
        }

        var result = await _client.ValidateCountryAsync(new Dictionary<string, object>
        {
            ["code"] = country,
            ["follow_upward"] = true
        });

        if (!result.Valid)
        {
            return UnprocessableEntity(result);
        }

        return Ok(new Dictionary<string, object>
        {
            ["valid"] = true,
            ["message"] = "Country validated with follow_upward",
            ["payload"] = payload
        });
    }
}

